compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp

```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
+      auto accessorFunction = solver.getElementFor<GlobalAccessor>(
+          *this, Position::forOperation(callableOp), DFX::Resolution::REQUIRED);
+      LLVM_DEBUG({
+        llvm::dbgs() << "[simplify-globals] call: ";
+        callOp.print(llvm::dbgs(), solver.getAsmState());
+        llvm::dbgs() << "; ";
+        accessorFunction.print(llvm::dbgs(), solver.getAsmState());
+        llvm::dbgs() << "\n";
+      });
+      StateType newState = accessorFunction;
+      return DFX::inheritStateAndIndicateChange(getState(), newState);
+    }
+
+    // At this point the only unhandled operations are those with regions. To
+    // determine the current known number of global accessors, sum the accesses
+    // of the ops within this immediate region.
+    assert(op->getNumRegions() > 0 && "updating regionless global accessor");
+    StateType bestState = getBestState();
+    for (auto &region : op->getRegions()) {
+      for (auto &block : region) {
+        for (auto &containedOp : block) {
+          auto accessor = solver.getElementFor<GlobalAccessor>(
+              *this, Position::forOperation(&containedOp),
+              DFX::Resolution::REQUIRED);
+          bestState ^= accessor;
+        }
+      }
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "[simplify-globals] region op: ";
+      op->print(llvm::dbgs(), solver.getAsmState());
+      llvm::dbgs() << "; " << bestState;
+      llvm::dbgs() << "\n";
+    });
+    return DFX::inheritStateAndIndicateChange(getState(), bestState);
+  }
+
+  friend class DFX::Solver;
+};
+const char GlobalAccessor::ID = 0;
+
+class GlobalAccessorAnalysis {
+public:
+  explicit GlobalAccessorAnalysis(Operation *rootOp)
+      : explorer(rootOp, TraversalAction::RECURSE,
+                 OpPrintingFlags().elideLargeElementsAttrs().skipRegions()),
+        solver(explorer, allocator) {
+    // TODO: Use op traits or something to ignore the contents of executables
+    // and execution regions (they don't access globals). i.e.
+    //   explorer.setOpAction<IREE::Stream::ExecutableOp>(TraversalAction::IGNORE);
+    //   explorer.setOpAction<IREE::HAL::ExecutableOp>(TraversalAction::IGNORE);
+    explorer.initialize();
+
+    assert(rootOp->getNumRegions() == 1 && "expected module-like root op");
+
+    // Populate the list of globals to optimize. This is primarily to force the
+    // order in which globals are processed to be deterministic, reducing minor
+    // ordering variations in the resulting IR.
+    rootOp->walk([&](IREE::Util::GlobalOpInterface global) {
+      globalNames.push_back(global.getGlobalName());
+    });
+  }
+
+  AsmState &getAsmState() { return solver.getAsmState(); }
+  Explorer &getExplorer() { return explorer; }
+
+  // Runs analysis and populates the state cache.
+  // May fail if analysis cannot be completed due to unsupported or unknown IR.
+  LogicalResult run() {
+    // Initialize all potential accessors.
+    explorer.walk([&](Operation *op) {
```

Comment:
use explorer.forEachGlobal to avoid the full module walk - it already should have all uses of global symbols in the `uses` list and you can use `isIndirect` to indicate a pessimistic state immediately

Reply:
A comment and a couple questions: In addition to initializing the state for global.load/store, I'm also initializing it for things like external function calls, ops with Util_YieldPoint`, and indirect accessors. For external function calls I'm guessing I can just inspect the call graph, but I'm a bit unclear on how indirect accessors work nor how to handle yield points.

For indirect accessors, what I'm assuming now is that if I have an indirect accessor, I can't move any direct loads/stores past it, but I can still optimize around it, i.e.
```
// Some direct loads/stores I can optimize
util.global.load.indirect
// Some direct loads/stores I can optimize
```
I'm not checking whether the indirect load was loading/storing the same global as the nearby direct loads/stores (as I thought indirect more or less implied that we can't know which global it's accessing), but `GlobalInfo` seems to indicate that the explorer is able to do some reasoning about them? i.e. https://github.com/openxla/iree/blob/c27ed41cb17e8e8ce282cf9b838b0af1dd48c540/compiler/src/iree/compiler/Dialect/Util/Analysis/Explorer.h#L129

My expectation was that even if a global should be considered volatile due to indirect accesses, the indirect load and store ops are still the only way to interact with them, and thus we can still optimize other loads of that global. Is this understanding correct? Or do I have my understanding of indirect accessors mixed up (I kind of just pieced it together from code/comments).

If we say we'd rather not bother trying to ever optimize in the presence of indirect loads/stores that's fine and makes this easier, otherwise I'm not sure how to avoid the full walk because the explorer doesn't track these accessors.

My second question is w.r.t. yield points. I've been assuming we can't push loads/stores past, say, a `vm.break` because the user needs to be able to inspect the proper module state at that point in the code. As a result, I don't know how to avoid this walk for the same reason because I need to find all of the yield points throughout the module. Is there a reasonable way to populate a set of yield points on initialization of the explorer? Else I could let the initialization of yield points happen as an update step of the analysis, but that seems like it would introduce more IR walking than this.

Reply:
Should've read the code closer, I see that the explorer is just looking for globals where their address is taken, i.e. `GlobalAddressOpInterface`. So the explorer never sees the indirect loads/stores. I see that `initializeInverseCallGraph` still does a full module walk though, could we reuse that module walk to populate a list of yield points/indirect loads?